src/engine/enginebuffer.cpp

```diff
@@ -740,6 +752,18 @@ void EngineBuffer::processTrackLocked(
     bool is_scratching = false;
     bool is_reverse = false;
 
+    if (!m_activeMacro.empty()) {
+        double framePos = m_filepos_play / kSamplesPerFrame;
+        MacroAction* nextAction = m_activeMacro.front();
+        double nextActionPos = nextAction->position;
+        // this method is called roughly every iBufferSize samples (double as often if you view frames)
+        // so use double that as the tolerance range
+        if (framePos > nextActionPos - iBufferSize && framePos < nextActionPos) {
+            slotControlSeekExact(m_activeMacro.front()->target * kSamplesPerFrame);
```

Comment:
You should avoid redundant calls to front()/pop() which involve memory barriers.